lib/std/date/gregorian.zig

```diff
@@ -0,0 +1,397 @@
+    try expectEqual(30, MonthT.sep.days(false));
+    try expectEqual(31, MonthT.oct.days(false));
+    try expectEqual(30, MonthT.nov.days(false));
+    try expectEqual(31, MonthT.dec.days(false));
+}
+
+pub fn is_leap(year: anytype) bool {
+    return if (@mod(year, 25) != 0)
+        year & (4 - 1) == 0
+    else
+        year & (16 - 1) == 0;
+}
+
+test is_leap {
+    try expectEqual(false, is_leap(2095));
+    try expectEqual(true, is_leap(2096));
+    try expectEqual(false, is_leap(2100));
+    try expectEqual(true, is_leap(2400));
+}
+
+pub fn daysBetweenYears(from: usize, to: usize) usize {
+    var res: usize = 0;
+    var i: usize = from;
+    while (i < to) : (i += 1) {
+        res += if (is_leap(i)) 366 else 365;
+    }
+    return res;
+}
+
+test daysBetweenYears {
+    try expectEqual(366, daysBetweenYears(2000, 2001));
+    try expectEqual(146_097, daysBetweenYears(0, 400));
+}
+
+/// The Gregorian calendar repeats every 400 years.
+const era = 400;
+const era_days: comptime_int = daysBetweenYears(0, 400); // 146_097
+
+/// Number of days between two consecutive March equinoxes
+const days_in_year = struct {
+    const actual = 365.2424;
+    // .0001 days per year of error.
+    const numerator = 1_461;
+    const denominator = 4;
+};
+
+/// Int type to represent all possible days between minimum `Year` and maximum `Year`.
+/// Rounded up to nearest power of 2 to meet unsigned division assertions.
+fn MakeEpochDays(comptime Year: type) type {
+    const year_range = (std.math.maxInt(Year) - std.math.minInt(Year)) * days_in_year.actual;
+    const required_int_bits = @ceil(std.math.log2(year_range));
+    const int_bits = std.math.ceilPowerOfTwoAssert(u16, @intFromFloat(required_int_bits));
+    return std.meta.Int(.signed, int_bits);
+}
+
+fn UIntFitting(to: comptime_int) type {
+    return IntFittingRange(0, to);
+}
+
+/// Finds minimum epoch shift that covers the range:
+/// [std.math.minInt(YearT), std.math.maxInt(YearT)]
+fn solveShift(comptime Year: type, epoch: comptime_int) !comptime_int {
+    const shift = std.math.maxInt(Year) / era + 1;
+
+    const L = era * shift;
+    const K = epoch + era_days * shift;
+    const EpochDays = MakeEpochDays(Year);
+    const max_day = (std.math.maxInt(EpochDays) - 3) / 4 - K;
+
+    if (@divFloor(max_day, days_in_year.numerator) - L + 1 <= std.math.maxInt(Year))
+        return error.Constraint; // if you hit this write a system of equations solver here to prove
+    // there are no possible values of `shift`
+
+    return shift;
+}
+
+test solveShift {
+    try expectEqual(82, try solveShift(i16, 719_468));
+    try expectEqual(5_368_710, try solveShift(i32, 719_468));
+    try expectEqual(23_058_430_092_136_940, try solveShift(i64, 719_468));
+}
+
+const std = @import("std");
+const IntFittingRange = std.math.IntFittingRange;
+const secs_per_day = std.time.s_per_day;
+const expectEqual = std.testing.expectEqual;
+const assert = std.debug.assert;
```

Comment:
These imports only go on the bottom on when the file is used as a struct instead of a namespace.

Reply:
Done.